torch_em/loss/wrapper.py

```diff
@@ -43,30 +44,72 @@ def forward(self, prediction, target, **kwargs):
 
 
 class ApplyMask:
-    def __call__(self, prediction, target, mask):
-        mask.requires_grad = False
+    def _crop(prediction, target, mask, channel_dim):
+        mask = mask.type(torch.bool)
+        pred = []
+        trgt = []
+        n_channels = prediction.shape[channel_dim]
+        # for now only support same mask for all channels
+        # to be able to stack/concat afterwards
+        # TODO: make this more efficient, and for different
```

Comment:
For the case where we have a mask of lower dimensionality this can be done much simpler (I am using numpy here for simplicity, but the same works with torch, I just checked):
```
x = np.random.rand(128, 128, 3)
y = np.random.rand(128, 128) > 0.5
z = x[y]
print(z.shape)
(8186, 3)  # 8186 is the number of elems in mask, i.e. it's not always the same number
```
So you just need to remove the singleton channel dimension in the mask, swap the channel dimension to the back in the prediction / target, apply the mask and bring the channel dimension back to the original pos.

I still need to think about the case where we have a full mask.

Reply:
True, I implement it. Thank you :)

Reply:
done